src/platform/PICO/uart/uart_pio.c

```diff
@@ -0,0 +1,374 @@
+ * and/or modify this software under the terms of the GNU General
+ * Public License as published by the Free Software Foundation,
+ * either version 3 of the License, or (at your option) any later
+ * version.
+ *
+ * Betaflight is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ *
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this software.
+ *
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "platform.h"
+
+#ifdef USE_UART
+
+#include "drivers/io.h"
+#include "drivers/serial.h"
+#include "drivers/serial_uart.h"
+#include "drivers/serial_impl.h"
+#include "drivers/serial_uart_impl.h"
+#include "hardware/irq.h"
+#include "hardware/pio.h"
+
+#include "serial_uart_pico.h"
+
+// PIOUARTs are used in order, starting from PIOUART0
+STATIC_ASSERT(SERIAL_PIOUART_COUNT == SERIAL_PIOUART_MAX, should_be_enforced_PIOUART_COUNT_IS_PIOUART_MAX);
+
+// The PIO block for software UARTs PIOUART0, PIOUART1
+static const PIO uartPio = PIO_INSTANCE(PIO_UART_INDEX);
+
+#define PIO_IRQ_INDEX(irqn) ((irqn) == PIO_IRQ_NUM(uartPio, 0) ? 0 : 1)
+
+typedef struct pioDetails_s {
+    irq_num_t irqn;
+    io_rw_32 *enableReg;
+    io_ro_32 *statusReg;
+    int rxPin;
+    int txPin;
+    uint16_t sm_rx; // sm number for rx (0..3)
+    uint16_t sm_tx; // sm number for tx (0..3)
+    uint32_t rx_intr_bit; // bit to check on interrupt enable and status registers for rx not empty
+    uint32_t tx_intr_bit; // bit to check on interrupt enable and status registers for tx not full
+} pioDetails_t;
+
+#if SERIAL_PIOUART_COUNT > 2
+#error USE_PIOUARTn only currently supported for n=0,1
+#endif
+
+// Store for details, catering (currently) for PIOUART0, PIOUART1
+static pioDetails_t uartPioDetails[SERIAL_PIOUART_COUNT];
+
+#define UART_PIO_DETAILS_IDX(id) ((id) - SERIAL_PORT_PIOUART_FIRST)
+#define UART_PIO_DETAILS_PTR(id) (&uartPioDetails[UART_PIO_DETAILS_IDX(id)])
+
+// Base for PIO pin counts (0 or 16)
+static int uartPioBase;
+
+static int txProgram_offset = -1;
+static int rxProgram_offset = -1;
+
+// Look-ups for bit to check on status register IRQ[01]_INTS by SM
+static const uint32_t rxnemptybit[4] = {
+    PIO_INTR_SM0_RXNEMPTY_BITS,
+    PIO_INTR_SM1_RXNEMPTY_BITS,
+    PIO_INTR_SM2_RXNEMPTY_BITS,
+    PIO_INTR_SM3_RXNEMPTY_BITS,
+};
+
+static const uint32_t txnfullbit[4] = {
+    PIO_INTR_SM0_TXNFULL_BITS,
+    PIO_INTR_SM1_TXNFULL_BITS,
+    PIO_INTR_SM2_TXNFULL_BITS,
+    PIO_INTR_SM3_TXNFULL_BITS,
+};
+
+// PIO-based UARTs. For now, hardwired to PIOUARTs 0,1 on PIO number UART_PIO_INDEX.
+const pioUartHardware_t pioUartHardware[SERIAL_PIOUART_COUNT] = {
+#ifdef USE_PIOUART0
+    {
+        .identifier = SERIAL_PORT_PIOUART0,
+        .irqn = PIO_IRQ_NUM(uartPio, 0),
```

Comment:
irqn handling is quite complicated. Maybe it would be easier to invert it - store `.irqn_index` here and apply `PIO_IRQ_NUM` where irqn is needed